grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceImplErrorTest.java

```diff
@@ -0,0 +1,319 @@
+        protected void configure(ServerBuilder sb) {
+            final ServerInterceptor corruptedServerInterceptor = new ServerInterceptor() {
+                @Override
+                public <I, O> Listener<I> interceptCall(ServerCall<I, O> call,
+                                                        Metadata headers,
+                                                        ServerCallHandler<I, O> next) {
+                    return next.startCall(new SimpleForwardingServerCall<I, O>(call) {
+                        @Override
+                        public void close(Status status, Metadata trailers) {
+                            super.close(status, corruptedMetadata);
+                        }
+                    }, headers);
+                }
+            };
+            configureGrpcService(sb, corruptedServerInterceptor);
+        }
+    };
+
+    // Normal case of #onError at Unary RPC, but metadata set in server interceptor is corrupted.
+    // Client cannot expect corrupted metadata is returned from server.
+    @Test
+    void clientUnaryCall2ForServerUsingCorruptedInterceptor() throws InterruptedException {
+        try (ClientRequestContextCaptor captor = Clients.newContextCaptor()) {
+            assertThatThrownBy(() -> {
+                final TestServiceBlockingStub client =
+                        GrpcClients.newClient("http://127.0.0.1:" + serverUsingCorruptedInterceptor.httpPort(),
+                                              TestServiceBlockingStub.class);
+                client.unaryCall2(REQUEST_MESSAGE);
+            }).satisfies(cause -> {
+                assertThat(Status.fromThrowable(cause).getCode()).isEqualTo(Status.INTERNAL.getCode());
+                assertThat(Status.trailersFromThrowable(cause)).satisfies(metadata -> {
+                    assertThat(isSameSizeAndNameAndValues(metadata, corruptedMetadata)).isFalse();
+                });
+            });
+
+            final RequestLog log = captor.get().log().whenComplete().join();
+            assertThat(log.responseStatus()).isEqualTo(HttpStatus.OK);
+            assertThat(log.responseTrailers().get(GrpcHeaderNames.GRPC_STATUS)).isNull();
+            assertThat(log.responseHeaders().get(GrpcHeaderNames.GRPC_STATUS)).satisfies(grpcStatus -> {
+                assertThat(grpcStatus).isNotNull();
+                assertThat(grpcStatus).isEqualTo(String.valueOf(Status.INTERNAL.getCode().value()));
+            });
+            assertThat(log.responseCause())
+                    .isInstanceOf(StatusRuntimeException.class)
+                    .satisfies(cause -> {
+                        assertThat(Status.fromThrowable(cause).getCode()).isEqualTo(Status.INTERNAL.getCode());
+                        assertThat(Status.trailersFromThrowable(cause)).satisfies(metadata -> {
+                            assertThat(isSameSizeAndNameAndValues(metadata, corruptedMetadata)).isFalse();
+                        });
+                    });
+        }
```

Comment:
I think we should be checking server-side `responseCause` instead of client-side.

```suggestion
final ServiceRequestContextCaptor serviceCaptor = server.requestContextCaptor();
assertThat(serviceCaptor.size()).isEqualTo(1);
final ServiceRequestContext serviceCtx = serviceCaptor.take();
assertThat(serviceCtx.log().ensureComplete().responseCause())
        .isInstanceOf(IndexOutOfBoundsException.class);
```

Reply:
Why do you think so?

From the implementor side of client,
we should test from the view point of client (mine) (JFYI: https://github.com/line/armeria/pull/4686#discussion_r1127457836).

But from the implementor of server side,
your suggestion is right.

As a conclusion,
should not we accept the both tests?

Reply:
> From the implementor side of client,
we should test from the view point of client (mine) (JFYI: https://github.com/line/armeria/pull/4686#discussion_r1127457836).

I think the original comment you mentioned is referring to checking headers/trailers which make sense.

The current code I mentioned above seems to be checking whether the correct `StatusRuntimeException` will be raised if `grpc-status = 13` is returned.

I think it's fine to check this, but I still think we need the check on `ServiceRequestContextCaptor` which is missing at the moment. Without the check on server-side, I would guess server-side owners will not be aware of why a response has failed if the cause is not shown in server-side `log.responseCause`.